File: src/dataset/cpp/nth-prime_exercism-chrisnp/openAI/ChatGPT4_nth_prime_v2.cpp
#include "nth_prime.h"
#include <stdexcept>
#include <vector>
#include <cmath>

unsigned long nth_prime::nth(const unsigned long &num) {
    if (num == 0) { throw std::domain_error("No zeroth prime"); }
    if (num == 1) { return 2; }

    std::vector<unsigned long> primes = {2};
    unsigned long candidate = 3;

    while (primes.size() < num) {
        bool is_prime = true;
        unsigned long limit = std::sqrt(candidate);
        for (const auto &prime : primes) {
            if (prime > limit) break;
            if (candidate % prime == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            primes.push_back(candidate);
        }
        candidate += 2; // Skip even numbers
    }
    return primes.back();
}

bool nth_prime::is_prime(const int &num) noexcept {
    if (num <= 3) return num > 1;
    if (num % 2 == 0 || num % 3 == 0) return false;
    for (int i = 5; i * i <= num; i += 6)
        if (num % i == 0 || num % (i + 2) == 0) 
            return false;
    return true;
}

File: src/dataset/python/nth-prime_exercism-fruit-in/openAI/ChatGPT4_nth_prime_v2.py
def prime(number):
    if number < 1:
        raise ValueError(r".+")

    if number == 1:
        return 2

    count = 1
    candidate = 3

    while count < number:
        if isprime(candidate):
            count += 1
        candidate += 2

    return candidate - 2


def isprime(number):
    if number < 2:
        return False
    if number in (2, 3):
        return True
    if number % 2 == 0 or number % 3 == 0:
        return False

    i = 5
    while i * i <= number:
        if number % i == 0 or number % (i + 2) == 0:
            return False
        i += 6

    return True